File: src/leetcode/python/hard/PopulatingNextRightPointersInEachNodeII_117.py
class TreeLinkNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
        self.next = None


class PopulatingNextRightPointersInEachNodeII_117(object):
    def connect(self, root):
        """
        :type root: TreeLinkNode
        :rtype: nothing
        """
        def next_node(linke_node):
            node = linke_node.next
            while node:
                if node.left: return node.left
                if node.right: return node.right
                node = node.next
            return None

        if not root:
            return

        parent, left_most = None, root
        while left_most:
            parent = left_most
            left_most = None
            while parent:
                if parent.left:
                    parent.left.next = parent.right if parent.right else next_node(parent)
                    if not left_most:
                        left_most = parent.left
                if parent.right:
                    parent.right.next = next_node(parent)
                    if not left_most:
                        left_most = parent.right
                parent = parent.next

File: src/leetcode/java/PopulatingNextRightPointersInEachNodeII_117.java
package leetcode.java.hard;

public class PopulatingNextRightPointersInEachNodeII_117 {
  public void connect(TreeLinkNode root) {
    if (root == null) return;

    TreeLinkNode leftMost = root;
    TreeLinkNode parent = null;
    while (leftMost != null) {
      parent = leftMost;
      leftMost = null;
      while (parent != null) {
        if (parent.left != null) {
          parent.left.next = (parent.right != null) ? parent.right : getNext(parent);
          if (leftMost == null) leftMost = parent.left;
        }
        if (parent.right != null) {
          parent.right.next = getNext(parent);
          if (leftMost == null) leftMost = parent.right;
        }
        parent = parent.next;
      }
    }
  }

  private TreeLinkNode getNext(TreeLinkNode parent) {
    TreeLinkNode node = parent.next;
    while (node != null) {
      if (node.left != null) return node.left;
      if (node.right != null) return node.right;
      node = node.next;
    }
    return null;
  }

  class TreeLinkNode {
    int val;
    TreeLinkNode left, right, next;
    TreeLinkNode(int x) { val = x; }
  }
}
